### ISSUE TITLE
[data types] Bloqs which work for multiple types

### ISSUE THREAD

**Issue by fdmalone** [quantumlib/Qualtran#637]
As part of #446 we will need to update the signatures of several arithmetic bloqs to allow for some generic behavour.

One way to do this is

``` python
class Add(bloq):
    dtype: QDType

   @dtype.validator(self):
        assert isinstance(dtype, (QInt, QUInt, ...))
```

This will happen after the migration work.

**Comment by mpharrigan** [comment]
I think I asked during the design review whether we'd need abstract qdtypes. Specifically: a named-but-not-instantiatable type to replace `(QInt, QUint, ...)` in the isinstance check (and appropriate for the attrs type annotation)

**Comment by fdmalone** [comment]
right, that is the plan once we sort of land on what bloqs work on which types.
